### PR TITLE
Fix for the handling of constant arrays and members of structs being marked nondet.

### PR DESCRIPTION
**Pull request by NlightNFotis**
This is a port of #1632, from goto-analyzer-develop to develop.

### CODE REVIEW

**Review by martin-cs — COMMENTED** [review]
Reviewed by @tautschnig @chrisr-diffblue and myself while a PR to goto-analyzer-develop : https://github.com/diffblue/cbmc/pull/1632

**Comment by tautschnig** [comment]
Would you mind making your first commit message a bit more verbose: "Fix for the constant arrays marked as nondet issue." -- which issue is it, and what changes are you making?

File: src/goto-instrument/nondet_static.cpp [hunk 1]
```diff
@@ -18,6 +18,7 @@ Date: November 2011
 #include <util/std_expr.h>
 #include <util/cprover_prefix.h>
 #include <util/prefix.h>
+#include <util/type.h>
```
Comment (tautschnig) [inline src/goto-instrument/nondet_static.cpp]: That include is unnecessary (it's dragged in via `std_expr.h` already).

File: regression/cbmc/no_nondet_static/test.desc [hunk 2]
```diff
@@ -0,0 +1,33 @@
+CORE
+main.c
+--nondet-static
+^VERIFICATION FAILED$
```
Comment (tautschnig) [inline regression/cbmc/no_nondet_static/test.desc:4]: Please also check SIGNALS and exit codes as other tests commonly do.

**Review by smowton — APPROVED** [review]

File: src/util/type.cpp [hunk 3]
```diff
@@ -34,3 +35,76 @@ bool is_number(const typet &type)
...
+    return type.get_bool(ID_C_constant);
+
+  // When we have a case like the following, we don't immediately
+  // see the struct t. Instead, we only get to see symbol t1, which
+  // we have to use the namespace to resolve to its definition:
+  // struct t { const int a; };
+  // struct t t1;
+  if(type.id() == ID_symbol)
+  {
+    const auto &subtype = ns.follow(type);
```
Comment (tautschnig) [inline src/util/type.cpp]: Nit pick: `subtype` is misleading, it's the same type, just resolved.

File: src/util/type.cpp [hunk 4]
```diff
@@ -34,3 +35,76 @@ bool is_number(const typet &type)
...
+    return has_constant_components(subtype);
+  }
+
+  // In a case like this, were we see an array (b[3] here), we know that
```
Comment (tautschnig) [inline src/util/type.cpp]: s/were/where/

File: src/util/type.cpp [hunk 5]
```diff
@@ -34,3 +35,76 @@ bool is_number(const typet &type)
...
+  // the array contains subtypes. We get the first one, and
+  // then resolve it to its  definition through the usage of the namespace.
+  // struct contains_constant_pointer { int x; int * const p; };
+  // struct contains_constant_pointer b[3] = { {23, &y}, {23, &y}, {23, &y} };
+  if(type.has_subtype())
+  {
+    const auto &subtype = type.subtype();
+    if(subtype.id() == ID_symbol)
```
Comment (tautschnig) [inline src/util/type.cpp]: You are no longer covering the case of `const int A[n]`, I'd say (also that case is missing from the test) -- which is really where the original PR started from if I remember correctly?

You just need to recurse in all cases, not just if the subtype is `ID_symbol`.

**Review by tautschnig — CHANGES_REQUESTED** [review]

File: src/goto-instrument/nondet_static.cpp (hunk 1 above)
Comment (NlightNFotis) [reply src/goto-instrument/nondet_static.cpp]: @tautschnig  There is some discussion that we shouldn't depend on transitive `#include`s. Do you think it definitely needs to go?

**Comment by NlightNFotis** [comment]
@tautschnig I addressed all the outstanding review comments. Can I have a new review at your earliest convenience? Thanks for your time.

File: src/goto-instrument/nondet_static.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-instrument/nondet_static.cpp]: Oh, I was not aware of that discussion. I was just very surprised to see this change.

File: src/util/type.cpp [hunk 6]
```diff
@@ -34,3 +35,80 @@ bool is_number(const typet &type)
...
+    {
+      const auto &new_subtype = ns.follow(to_symbol_type(subtype));
+      return has_constant_components(new_subtype);
+    }
+    else
+    {
+      return is_constant_or_has_constant_components(subtype, ns);
+    }
```
Comment (tautschnig) [inline src/util/type.cpp]: I think this code can now be simplified as the case line 88 would anyway cover `ID_symbol`?

Comment (NlightNFotis) [reply src/util/type.cpp]: Yes, you are right, I can simplify it.

File: src/util/type.cpp [hunk 7]
```diff
@@ -34,3 +35,72 @@ bool is_number(const typet &type)
...
+  // comprehensive test case for this, take a look at
+  // regression/cbmc/no_nondet_static/main.c
+
+  // const int a;
+  if(type.get_bool(ID_C_constant))
+    return true;
+
+  // Handles the case where we have a recursive data type definition
+  // like the linked list below.
+  // struct list { const int datum; struct list * next; };
```
Comment (chrisr-diffblue) [inline src/util/type.cpp]: This comment isn't strictly true - this case doesn't handle recursive structs, but I think what you meant is that this case is used to terminate the recursion in cases where the struct is a recursive structure. Might also be worth highlighting the fairly subtle distinction between this case (where for pointers you *always* return) and the case in lines 74/75 - where you only return if the type is a constant (and fall through to this case if not).

**Review by chrisr-diffblue — APPROVED** [review]
Looks good, and I'm happy for this to go in now.

**Review by tautschnig — APPROVED** [review]

**Comment by martin-cs** [comment]
Happy to commit but what is going on this AppVeyor?

**Comment by chrisr-diffblue** [comment]
I'm not sure what's happening with AppVeyor, but there was a suggestion from @reuk that it may be caused by some recent windows build work that he'd done, that seems to have got some PR's in this state. The failure is happening while AppVeyor is setting itself up so is not in anyway caused by this PR. I'd personally be happy for this to be committed even with AppVeyor being unhappy.

**Comment by chrisr-diffblue** [comment]
For reference, another AppVeyor job for a completely different PR, https://ci.appveyor.com/project/diffblue/cbmc/build/1.0.5894 shows the same failure mode.

**Comment by NlightNFotis** [comment]
After discussion with @thk123 the AppVeyor failure is due to a known issue with the Windows build, which is resolved now on develop. I will rebase against develop now, and I expect all of the CI builds to pass.

**Pull request merged**